ClientRuntimes/CSharp/Microsoft.Rest.ClientRuntime/BatchDelegatingHandler.cs

```diff
@@ -0,0 +1,185 @@
+        /// </summary>
+        /// <param name="method">the type of method that the Batch operation is</param>
+        /// <param name="url">the URL that the Batch operation needs to hit</param>
+        public BatchDelegatingHandler(HttpMethod method, Uri url)
+        {
+            this.batchMethod = method;
+            this.batchURL = url;
+            this.Reset();
+        }
+
+        /// <summary>
+        /// Resets the batch handler to accept a new set of batch requests.
+        /// </summary>
+        public void Reset()
+        {
+            this.requests = new List<HttpRequestMessage>();
+            this.responses = null;
+            this.batchCompleted = new ManualResetEvent(false);
+        }
+
+        /// <summary>
+        /// Send all the queued up requests as a single batch request to the inner handler.
+        /// </summary>
+        /// <param name="cancellationToken">Cancellation token for the batch request</param>
+        /// <param name="boundary">Optional string which will be used as the encapsulation boundary between each request in multipart content; if not specified, a random GUID will be used</param>
+        /// <returns>Task for the batch</returns>
+        public async Task IssueBatch(CancellationToken cancellationToken = default(CancellationToken), string boundary = null)
+        {
+            // There must be at least one request to issue
+            if (this.requests.Count == 0)
+            {
+                throw new Exception("No queued requests to issue.");
+            }
+
+            // Sanity check. There must not be any queued responses from the server.
+            if (this.responses != null)
+            {
+                throw new Exception("This batch handler instance has been reused without calling reset first.");
+            }
+
+            // Create a request to the batch URL
+            HttpRequestMessage batchRequest = new HttpRequestMessage(this.batchMethod, this.batchURL);
+
+            // Create the boundary
+            if (string.IsNullOrEmpty(boundary))
+            {
+                boundary = "batch_" + Guid.NewGuid().ToString();
+            }
+
+            // Add a multipart content to the body of the request
+            MultipartContent batchContent = new MultipartContent("mixed", boundary);
+            batchRequest.Content = batchContent;
+
+            // add each request to this batch request
+            foreach (HttpRequestMessage message in this.requests)
+            {
+                batchContent.Add(new HttpMessageContent(message));
+            }
+
+            // send the batch request to the inner handler
+            HttpResponseMessage batchResponse = await base.SendAsync(batchRequest, cancellationToken);
+
+            // throw exceptions if the response is unexpected
+            if (batchResponse == null)
+            {
+                throw new Exception("Batch request got back a null response.");
```

Comment:
Exception feels like the wrong exception type - seems like this should be some kind of Http or Web Exception